### ISSUE TITLE
AES-SIV EVP_DecryptUpdate() fails without setting error

### ISSUE THREAD

**Issue by petere** [openssl/openssl#18440]
(openssl-3.0.3)

I'm trying out AES-SIV. I'm using the test program below to try to decrypt the test vectors from RFC 5297 (test program also available at https://github.com/petere/test-openssl-aes-siv). The call to `EVP_DecryptUpdate()` (the second one, not the one for the AD) fails, but no error is set, so the output from the program is:

```
$ ./test-openssl-aes-siv
EVP_DecryptUpdate: (null) (0)
```
So, I'm probably doing something wrong (a hint would be welcome), but my point is that if the function fails it should set an error. My implementation of *printf apparently prints "(null)", but in other circumstances this could also crash, which is bad.

------

I traced the calls like this: `EVP_DecryptUpdate` ... `siv_cipher` ... `ossl_siv128_decrypt`, and there it somehow fails, but no error is set, and the code is so terse that I cannot guess the reason.

------

```c
#include <stdlib.h>

#include <openssl/err.h>
#include <openssl/evp.h>

unsigned char key[] = {
	0xff, 0xfe, 0xfd, 0xfc, 0xfb, 0xfa, 0xf9, 0xf8, 0xf7, 0xf6, 0xf5, 0xf4, 0xf3, 0xf2, 0xf1, 0xf0,
	0xf0, 0xf1, 0xf2, 0xf3, 0xf4, 0xf5, 0xf6, 0xf7, 0xf8, 0xf9, 0xfa, 0xfb, 0xfc, 0xfd, 0xfe, 0xff,
};

unsigned char ad[] = {
	0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
	0x20, 0x21, 0x22, 0x23, 0x24, 0x25, 0x26, 0x27,
};

unsigned char ciphertext[] = {
	0x40, 0xc0, 0x2b, 0x96, 0x90, 0xc4, 0xdc, 0x04, 0xda, 0xef, 0x7f, 0x6a, 0xfe, 0x5c,
};

static void
test1(void)
{
	EVP_CIPHER *cipher = EVP_CIPHER_fetch(NULL, "AES-128-SIV", NULL);
	if (!cipher)
	{
		fprintf(stderr, "EVP_CIPHER_fetch: %s\n",
			ERR_reason_error_string(ERR_get_error()));
		exit(1);
	}

	EVP_CIPHER_CTX *evp_cipher_ctx = EVP_CIPHER_CTX_new();

	if (!EVP_DecryptInit_ex(evp_cipher_ctx, cipher, NULL, NULL, NULL))
	{
		fprintf(stderr, "EVP_DecryptInit_ex: %s\n",
			ERR_reason_error_string(ERR_get_error()));
		exit(1);
	}

	/* key */
	if (!EVP_DecryptInit_ex(evp_cipher_ctx, NULL, NULL, key, NULL))
	{
		fprintf(stderr, "EVP_DecryptInit_ex#2: %s\n",
			ERR_reason_error_string(ERR_get_error()));
		exit(1);
	}

	int decrlen;

	/* AD */
	if (!EVP_DecryptUpdate(evp_cipher_ctx, NULL, &decrlen, ad, sizeof(ad)))
	{
		fprintf(stderr, "EVP_DecryptUpdate[AD]: %s\n",
			ERR_reason_error_string(ERR_get_error()));
		exit(1);
	}

	unsigned char outbuf[1024];
	if (!EVP_DecryptUpdate(evp_cipher_ctx, outbuf, &decrlen, ciphertext, sizeof(ciphertext)))
	{
		fprintf(stderr, "EVP_DecryptUpdate: %s (%lu)\n",
			ERR_reason_error_string(ERR_get_error()), ERR_get_error());
		exit(1);
	}

	int decrlen2;
	if (!EVP_DecryptFinal_ex(evp_cipher_ctx, outbuf + decrlen, &decrlen2))
	{
		fprintf(stderr, "EVP_DecryptFinal_ex: %s\n",
			ERR_reason_error_string(ERR_get_error()));
		exit(1);
	}
	decrlen += decrlen2;

	printf("plaintext = ");
	for (int i = 0; i < decrlen; i++)
		printf("0x%02x", outbuf[i]);
	printf("\n");
}

int
main(void)
{
	test1();

	return 0;
}
```

**Comment by slontis** [comment]
Decryption requires the tag to be set...
EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, tag_len, tag)

There is already code in the tests that does this operation on the test vector you have...
A cutdown version for decryption can use a file like this...

Add a file siv.txt that contains
````
Title = RFC5297 AES-SIV

Cipher = aes-128-siv
Operation = DECRYPT
Key = fffefdfcfbfaf9f8f7f6f5f4f3f2f1f0f0f1f2f3f4f5f6f7f8f9fafbfcfdfeff
AAD = 101112131415161718191a1b1c1d1e1f2021222324252627
Tag = 85632d07c6e8f37f950acd320a2ecc93
Plaintext =  112233445566778899aabbccddee
Ciphertext = 40c02b9690c4dc04daef7f6afe5c

````

And then run
````
./test/evp_test siv.txt
````

Step into cipher_test_enc() to see what functions it calls to do this operation.

**Comment by t8m** [comment]
Error codes on the error queue are optional. Sometimes they are raised sometimes not. If ERR_get_error() returns 0, there is no error in the error queue. Calling ERR_reason_error_string(0) returns NULL. Also please note that calling ERR_get_error() removes the returned error from the queue, so calling it again will return another error from the queue if there is any.

**Comment by petere** [comment]
Thanks for the feedback, I have got it working now. (I have updated my test repository with a full encrypt and decrypt test case, for posterity.)

A couple of observations:

- The term "tag" does not appear in RFC 5297 (AES-SIV). So the requirement in the OpenSSL API to set or get a tag is completely non-obvious to someone reading that specification. I suppose the term comes from AES-GCM-SIV. Some additional documentation would be useful to clarify this.

- Similarly, per RFC 5292, the output of the encryption operation is the generated initialization vector concatenated with the ciphertext. So I was assuming that a sequence of `EVP_Encrypt*` calls would produce that output. The fact that you need to stitch the tag and the encrypt output together yourself to get the output that the spec wants is again not obvious and should be documented.

- It was said here that getting error codes from `EVP_DecryptUpdate()` is optional, and I accept that, but I think it could be a bit better anyway. Forgetting to set the tag was obviously my mistake, but in my experimentation, I have also found that for example setting a wrong tag will lead to a failure of `EVP_DecryptUpdate()` without error code. So in the full application context, a user feeding in invalid, corrupted, unauthenticated, or wrong data for decryption will just result in a failure without explanation (such as "invalid data", "message authentication failed" etc.). This is not the experience that I have with other ciphers in OpenSSL.

- The documentation says that failure to authenticate a ciphertext (or tag or whatever) would report failure in `EVP_DecryptFinal()`, but I have found that not to be the case. As mentioned above, it fails in `EVP_DecryptUpdate()`. This is not such a big deal overall, but the documentation is very specific about that, so it would be good if it were also correct.

- Not sure if this is standard behavior for all ciphers or just something specific to AES-SIV: Passing a negative input length (`inl`) to `EVP_DecryptUpdate()` leads to a segmentation fault. Obviously, one can avoid that, but it seems to me that OpenSSL could support a little more robustness here. The circumstance here is that if my application provides the RFC-5297-style decryption input of "tag || ciphertext", then I naively pass `inputlen - taglen` as `inl` to `EVP_DecryptUpdate()`, which will work most of the time but will crash if something is passing me input shorter than `taglen`. Again, I can and should check that myself in my code, but it would be helpful if OpenSSL provided a well-defined error instead of crashing.

**Comment by t8m** [comment]
> * Not sure if this is standard behavior for all ciphers or just something specific to AES-SIV

You cannot really depend on openssl checking all the arguments of every API call for sanity. Yes, for new APIs we usually try to ensure that but EVP_Encrypt/DecryptUpdate is a very old one.

**Comment by tmshort** [comment]
The AES-SIV code was based on Daniel Franke's AES-SIV (https://github.com/dfoxfranke/libaes_siv).

The term "tag" is generic, and comes from other AEADs (OpenSSL does not yet support AES-GCM-SIV, defined in RFC8452, which came after AES-SIV, so no, the term tag did not come form AES-GCM-SIV).

The "tag" is the equivalent of the SIV *synthetic initialization vector" of RFC5297, that is part of the output of AES-SIV encryption. It is separate to work within the OpenSSL frameworks, and to allow the plain text and the cipher text buffers to be the same size.

In other words, it has similar behavior to the other AEADs within OpenSSL, except that it can only accept a single input of cipher text.

I read (actually, I wrote) the documentation, and yes, it would be better to indicate the differences from the RFC output and OpenSSL.

In the case of an error, `EVP_DecryptFinal()` will report an error. It just so happens that `EVP_DecryptUpdate()` does as well. So, if you were to have ignored the return value of `EVP_DecryptUpdate()`, `EVP_DecryptFinal()` will still report an error.

**Comment by tmshort** [comment]
In an unrelated, but somewhat related, note; I have started to look at implementing AES-GCM-SIV, but with 3.0, the crypto architecture changed quite a bit (and my day job got in the way), so it's taking some extra time.

**Comment by slontis** [comment]
> I have started to look at implementing AES-GCM-SIV, but with 3.0, the crypto architecture changed quite a bit (and my day job got in the way), so it's taking some extra time.

I did start having a look at this also quite a while ago.. I think it really needs a separate AEAD interface.

**Comment by kaduk** [comment]
> I think it really needs a separate AEAD interface.

I remember seeing that BoringSSL has an `EVP_AEAD` family of APIs but I haven't actually looked at them at all.

**Comment by t8m** [comment]
Hmm, but does it mean also adding a new provider API?

**Comment by mattcaswell** [comment]
> Hmm, but does it mean also adding a new provider API?

Yeah - probably. This has been on the TODO list for a loooonnnnnggg time. I recall having conversations about an AEAD API many years ago.

**Comment by paulidale** [comment]
I'm not sure this should have been automatically closed.  Hence reopening.

I do think a separate AEAD API is a good idea -- even now.

**Comment by petere** [comment]
Thanks for updating the documentation.

I'll leave you to figure out if you want a new API. ;-)

**Comment by t8m** [comment]
> 1. If GHash were better exposed

Can you be please more concrete what would be needed?

**Comment by tmshort** [comment]
> Can you be please more concrete what would be needed?

Make GHASH a publicly available hash algorithm... that being said, I figured out how to use it...

**Comment by t8m** [comment]
I do not think this is really actionable issue anymore.
